### PR TITLE
Picture in Picture improvements

### PR DESCRIPTION
**Pull request by ipavlidakis**
### 🎯 Goal

Make PiP stable.

### 🛠 Implementation
Currently supports:
- Always trigger PiP if a valid track exists
- Changing tracks while in PiP(including screensharing)
- Update tracks size while in PiP
- Bypasses the videoRendererFactory that causes issues fetching the related view and instead uses a placeholder view in order to present PiP. That causes a glitch when coming back with which we can either live or we need to invest more time removing the videoRendererFactory.

### 🎨 Showcase

_Add relevant screenshots and/or videos/gifs to easily see what this PR changes, if applicable._

|  Before  |  After  |
| -------- | ------- |
|  `img`   |  `img`  |

### 🧪 Manual Testing Notes

_Explain how this change can be tested manually, if applicable._

### ☑️ Contributor Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should receive manual QA
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (Docusaurus, tutorial, CMS)

### 🎁 Meme

_Provide a funny gif or image that relates to your work on this pull request. (Optional)_

### CODE REVIEW

File: Sources/StreamVideoSwiftUI/CallViewModel.swift [hunk 1]
```diff
@@ -140,6 +139,15 @@ open class CallViewModel: ObservableObject {
...
                 lastLayoutChange = Date()
+
+                let participants = call?.state.participants ?? []
+
+                guard
+                    participantsLayout == .grid,
+                    participants.count < 6,
+                    call?.state.screenSharingSession == nil
+                else { return }
+                participants.forEach { changeTrackVisibility(for: $0, isVisible: true) }
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallViewModel.swift:150]: Why do we need this?

File: Sources/StreamVideoSwiftUI/CallViewModel.swift [hunk 2]
```diff
@@ -164,15 +172,8 @@ open class CallViewModel: ObservableObject {
...
-    private lazy var pipTrackSelectionUtils = PiPTrackSelectionUtils()
-    
-    private var pipVideoRenderer: VideoRenderer? {
-        pipTrackSelectionUtils.pipVideoRenderer(
-            from: callParticipants,
-            currentSessionId: call?.state.sessionId
-        )
-    }
+
+    private var _pictureInPictureAdapter: Any?
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallViewModel.swift:176]: Maybe better to keep the type available on <iOS 15, to avoid using `Any`.

File: Sources/StreamVideoSwiftUI/CallViewModel.swift [hunk 3]
```diff
@@ -202,7 +203,15 @@ open class CallViewModel: ObservableObject {
         self.callSettings = callSettings ?? CallSettings()
         self.localCallSettingsChange = callSettings != nil
         self.subscribeToCallEvents()
-        self.subscribeForAppLifecycleEvents()
+//        self.subscribeForAppLifecycleEvents()
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallViewModel.swift:206]: We should delete this if we don't use it anymore.

File: Sources/StreamVideoSwiftUI/CallViewModel.swift [hunk 4]
```diff
@@ -202,7 +203,15 @@ open class CallViewModel: ObservableObject {
...
+
+        if #available(iOS 15.0, *) {
+            let adapter = StreamPictureInPictureAdapter.shared
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallViewModel.swift:209]: Maybe pass it via `Utils`, instead of having it shared?

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import AVKit
+import Foundation
+import StreamVideo
+import StreamWebRTC
+
+@available(iOS 15.0, *)
+final class StreamAVPictureInPictureVideoCallViewController: AVPictureInPictureVideoCallViewController {
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift:11]: The name is a bit too long, maybe we can do `StreamPiPCallViewController` or something?

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift [hunk 6]
```diff
@@ -0,0 +1,68 @@
+//
+//  StreamPictureInPIctureView.swift
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift:2]: Stream header.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import SwiftUI
+
+struct StreamPictureInPictureView: UIViewRepresentable {
+
+    func makeUIView(context: Context) -> UIView {
+        let view = UIView()
+        view.backgroundColor = .clear
+        if #available(iOS 15.0, *) {
+            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+                StreamPictureInPictureAdapter.shared.sourceView = view
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift:18]: same comment about `shared`, we should access it via our DI.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    func body(content: Content) -> some View {
+        content
+            .background(StreamPictureInPictureView())
+    }
+}
+
+extension View {
+
+    @ViewBuilder
+    public func enablePictureInPicture() -> some View {
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift:44]: Nice 👍 We should also document this probably.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        self.modifier(PictureInPictureModifier())
+    }
+}
+
+import UIKit
+
+extension UIApplication {
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift:51]: Why do we need this gorgeous code? 😅

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureController.swift [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            pictureInPictureController?.isPictureInPictureActive == true
+        else {
+            log.debug("Picture in Picture trackActivator is disabled!")
+            ensureActiveTrackIsEnabledCancellable?.cancel()
+            ensureActiveTrackIsEnabledCancellable = nil
+            return
+        }
+
+        log.debug("Picture in Picture trackActivator is active!")
+        ensureActiveTrackIsEnabledCancellable = Timer
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureController.swift:127]: why is this needed? Does it try to overwrite something that was done in the LLC?

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureTrackAdapter.swift [hunk 11]
```diff
@@ -0,0 +1,55 @@
+//
+//  StreamPictureInPictureTrackAdapter.swift
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureTrackAdapter.swift:2]: Stream header.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+        if #available(iOS 14.0, *) {
+            if contentView.sampleBufferDisplayLayer.requiresFlushToResumeDecoding == true {
+                contentView.sampleBufferDisplayLayer.flush()
+                log.debug("→ Display layer for track:\(trackId) flushed ✅")
+            }
+        }
+        if contentView.sampleBufferDisplayLayer.isReadyForMoreMediaData == true {
+            contentView.sampleBufferDisplayLayer.enqueue(buffer)
+//            log.debug("→ Frame for track:\(trackId) enqueued on picture in picture buffer ✅")
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift:113]: should this be uncommented?

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        }
+    }
+
+    private func startPictureInPicture(
+        for track: RTCVideoTrack?,
+        on window: UIWindow?
+    ) {
+        guard window != nil, let track else { return }
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift:121]: Why do we need the window, could you explain?

**Review by martinmitrevski — COMMENTED** [review]
Looks great so far 👌  Added some comments, but nothing major. What's really important is to fix the bug that happens when someone starts screensharing - the local track is lost again.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift (hunk 5 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift]: I don't like the PiP abbreviation and in any ways this is an internal object.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift (hunk 9 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPIctureView.swift]: We didn't at the end :P I hated it and i'm very happy i got to remove it haha

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureController.swift (hunk 10 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureController.swift]: This is to ensure that the Picture in Picture track will be enabled. As Picture in Picture isn't a responsibility of LLC I don't think that should be handled there. For now LLC pauses all tracks and after that doesn't have to worry about anything else.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift (hunk 12 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift]: It's not needed, it was only for me.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift (hunk 13 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift:192]: I have added comments now explaining those files, but to reiterate here, if the Picture in Picture window isn't visible, then we aren't consuming video frames, resulting in saving some computing operations.

File: Sources/StreamVideoSwiftUI/CallViewModel.swift (hunk 1 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/CallViewModel.swift:148]: I think i can remove that as it was a try to fix the disappearing track but i'm unsure it made any difference. Let me check

Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/CallViewModel.swift]: Removed and provided the fix in #263

File: Sources/StreamVideoSwiftUI/Utils.swift [hunk 14]
```diff
@@ -17,13 +18,9 @@ public class Utils {
 
 // MARK: - Utils + Default
 
-public extension Utils {
-    static var `default`: Utils = .init()
```
Comment (ipavlidakis) [inline Sources/StreamVideoSwiftUI/Utils.swift:21]: Removing it as there is always supposed to be a single instance of Utils and this one is breaking this "contract".

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    /// - Returns: A transformed CVPixelBuffer or nil if transformation fails.
+    func transform(
+        _ source: RTCI420Buffer,
+        targetSize: CGSize
+    ) -> CVPixelBuffer? {
+        let sourceSize = CGSize(width: Int(source.width), height: Int(source.height))
+
+        guard
+            requiresResize,
+            let resizedSource = resize(source, to: resizeSize(sourceSize, toFitWithin: targetSize))
```
Comment (ipavlidakis) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift:29]: Resizing using WebRTC's downscaling methods is actually more performant than rendering a very big (in dimensions) RTCVideoFrame. With this in place we can render tracks irregardless of their size without skyrocketing performance (that applies to both screensharing or not).

File: Sources/StreamVideoSwiftUI/Utils.swift [hunk 16]
```diff
@@ -9,6 +9,7 @@ public class Utils {
     public var userListProvider: UserListProvider
     public var callSoundsPlayer = CallSoundsPlayer()
     internal var videoRendererFactory = VideoRendererFactory()
+    internal let pictureInPictureAdapter = StreamPictureInPictureAdapter()
```
Comment (ipavlidakis) [inline Sources/StreamVideoSwiftUI/Utils.swift:12]: I'm thinking to actually make this a separate injection key from utils. The reason being that utils, can be created and we may end up with more than one instances of the adapter. WDYT?

Comment (martinmitrevski) [reply Sources/StreamVideoSwiftUI/Utils.swift:12]: yes, that's a good idea 👍

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    
+    /// The closure to call whenever the picture-in-picture window size changes.
+    var onSizeUpdate: ((CGSize) -> Void)? { get set }
+
+    /// The track that will be rendered on picture-in-picture window.
+    var track: RTCVideoTrack? { get set }
+
+    /// The preferred size for the picture-in-picture window.
+    /// - Important: This should **always** be greater to ``CGSize.zero``. If not, iOS throws
+    /// a cryptic error with content `PGPegasus code:-1003`
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift:21]: that's what that pegasus code meant 😁

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift [hunk 18]
```diff
@@ -0,0 +1,214 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+import StreamVideo
+import StreamWebRTC
+
+/// `StreamBufferTransformer` is a struct that provides methods for transforming RTCI420Buffer to
+/// CVPixelBuffer. while performing downsampling when necessary.
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift:10]: nit: comma instead of dot here

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift (hunk 15 above)
Comment (martinmitrevski) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift:29]: nice find 👍

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+struct StreamBufferTransformer {
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamBufferTransformer.swift:11]: Looks a lot nice now 👍

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureAdapter.swift [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+    /// Whenever participants change we update our internal state in order to always have the correct track
+    /// on picture-in-picture.
+    @MainActor
+    private func didUpdate(_ participants: [CallParticipant]) {
+        let sessionId = call?.state.sessionId
+        let otherParticipants = participants.filter { $0.sessionId != sessionId }
+
+        if let session = call?.state.screenSharingSession, call?.state.isCurrentUserScreensharing == false, let track = session.track {
+            pictureInPictureController?.track = track
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureAdapter.swift:76]: Now it worked with screensharing well 👍

File: docusaurus/docs/iOS/06-advanced/05-picture-in-picture.mdx [hunk 21]
```diff
@@ -7,6 +7,10 @@ Picture-in-Picture (PiP) is an essential part of video calls on mobile. It provi
 
 If you use our view components, as soon as you enter background, the native iOS PiP view will appear. It will show the first participant (based on the participant sorting criteria), that is not the current user. If a user is screen sharing, the screen sharing track would be shown instead.
 
+### Toggling Picture-in-Picture support
+
+You have control over when the Picture-in-Picture is enabled by accessing the `isPictureInPictureEnabled` property on the `CallViewModel`. By default, the value is set to `true`.
```
Comment (martinmitrevski) [inline docusaurus/docs/iOS/06-advanced/05-picture-in-picture.mdx:12]: Can we also mention the modifier in case you build your own view?

**Review by martinmitrevski — APPROVED** [review]
Awesome work 👏  The code is now reorganized nicely and it works really well, great job! 
Added few small things, but no need for a re-review.

File: Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift (hunk 17 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift:21]: Yeap and it was VERY difficult to figure it out. So I thought to add a note here so we can track it down (or anyone else searching the internet)

Comment (martinmitrevski) [reply Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamAVPictureInPictureVideoCallViewController.swift:21]: ChatGPT will be able to answer it in 2 years 😄 cc sam
